python/examples/music/client.py

```diff
@@ -0,0 +1,378 @@
+            albumName=track["album"]["name"],
+            uri=track["uri"],
+        )
+        for track in resultItems
+    ]
+
+    return tracks
+
+
+async def get_tracks_from_search(query: str, context: ClientContext) -> list[SimplifiedTrackInfo]:
+    results = await context.service.search(q=query, type='track', limit=50, offset=0)
+    tracks: list[SimplifiedTrackInfo] = []
+    while results:
+        tracks.extend(get_tracks_from_result_list(resultItems=results['tracks']['items']))
+
+        if results['next']:
+            results = await context.service.next(results)
+        else:
+            results = None
+
+    return tracks
+
+
+async def get_tracks_with_genres(
+    tracks: list[SimplifiedTrackInfo], context: ClientContext
+) -> list[SimplifiedTrackInfo]:
+    unique_artist_ids: list[str] = list(set([a for a in track.artistUris]))  # type: ignore
+    genre_lookup: dict[str, list[str]] = {}
+    for artist_id in unique_artist_ids:
+        artist = await context.service.artist(artist_id)
+        genre_lookup[artist_id] = [g.casefold() for g in artist["genres"]]
+    for track in tracks:
+        track_genres = set()
+        for artist_id in track.artistUris:
+            track_genres += set(genre_lookup[artist_id])  # type: ignore
+        track.genres = list(track_genres)
+
+    return tracks
+
+
+def print_tracks(tracks: list[SimplifiedTrackInfo]):
+    for track in tracks:
+        print(f" {track.name}")
+        print(f"   Artists: {", ".join(track.artistNames)}")
+        print(f"   Album: {track.albumName}")
+
+
+def update_track_list_and_print(tracks: list[SimplifiedTrackInfo], context: ClientContext):
+    print_tracks(tracks)
+    context.currentTrackList = tracks
+
+
+async def get_current_users_playlists(context: ClientContext) -> list[SimplifiedPlaylistInfo]:
+    results = await context.service.current_user_playlists(limit=50)
+    playlists: list[SimplifiedPlaylistInfo] = []
+    while results:
+        playlists.extend(
+            [SimplifiedPlaylistInfo(name=curr_list["name"], id=curr_list["id"]) for curr_list in results['items']]
+        )
+
+        if results['next']:
+            results = await context.service.next(results)
+        else:
+            results = None
+
+    return playlists
+
+
+async def print_status(context: ClientContext):
+    state = await context.service.current_playback()
+    if not state:
+        print("Nothing playing according to Spotify")
+    await list_available_devices(context)
+
+
+async def list_available_devices(context: ClientContext):
+    devices = await context.service.devices()
+    for device in devices["devices"]:
+        if device["is_active"]:
+            print(F"Active Device {device['name']} of type {device['type']}")
+        else:
+            print(f"Device {device['name']} of type {device['type']} is available")
+
+
+async def handle_call(action: PlayerAction, context: ClientContext):
+    match action["actionName"]:
```

Comment:
Something we could do is use a class with a bunch of methods with `hasattr`. I've suggested we could do something similar with a plain object in the TypeScript implementation too.

For now it is probably fine as-is. I wouldn't rewrite everything unless you think you can do it quickly.

Reply:
I've been looking for a way to verify type signatures at runtime. As far as I can tell, the TypedDicts are just dicts at runtime so we cannot "assert x is T" or "cast x as T" to ensure the right properties are present, Let me know if you find a way to do this.